src/epics.c

```diff
@@ -0,0 +1,551 @@
+/**
+ * collectd - src/epics.c
+ * Copyright (C) 2022       Matwey V. Kornilov
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Authors:
+ *   Matwey V. Kornilov <matwey.kornilov at gmail.com>
+ **/
+
+#include "collectd.h"
+#include "plugin.h"
+#include "utils/common/common.h"
+
+#include <cadef.h>
+
+struct pv {
+  char *name;
+
+  chtype ch_type;
+  chid id;
+  evid eid;
+
+  union {
+    value_list_t value;
+    struct {
+      char *value;
+      size_t capacity;
+    } label;
+  };
+
+  int is_active : 1;
+  int is_label : 1;
+};
+
+static struct {
+  struct pv *pvs;
+  int pvs_num;
+
+  pthread_mutex_t lock;
+  pthread_t thread_id;
+
+  int thread_loop : 1;
+} epics_plugin = {.pvs = NULL,
+                  .pvs_num = 0,
+                  .lock = PTHREAD_MUTEX_INITIALIZER,
+                  .thread_loop = 0};
+
+static void free_pvs() {
+  for (; epics_plugin.pvs_num > 0; --epics_plugin.pvs_num) {
+    struct pv *p = &epics_plugin.pvs[epics_plugin.pvs_num - 1];
+
+    if (p->is_label) {
+      free(p->label.value);
+    } else {
+      free(p->value.values);
+    }
+    free(p->name);
+  }
+
+  free(epics_plugin.pvs);
+  epics_plugin.pvs = NULL;
+}
+
+static int printf_handler(const char *pformat, va_list args) {
+#if COLLECT_DEBUG
+  char msg[1024] = ""; // Size inherits from plugin_log()
+  int status = 0;
+
+  status = vsnprintf(msg, sizeof(msg), pformat, args);
+
+  if (status < 0) {
+    return status;
+  }
+
+  msg[strcspn(msg, "\r\n")] = '\0';
+  plugin_log(LOG_DEBUG, "%s", msg);
+
+  return status;
+#else
+  return 0;
+#endif
+}
+
+static void handle_var_event(struct pv *p, evargs args) {
+  const data_set_t *ds = plugin_get_ds(p->value.type);
+  const size_t values_len = p->value.values_len;
+  value_t *values = p->value.values;
+
+  if (ds == NULL) {
+    ERROR("epics plugin: Unknown type \"%s\" for channel \"%s\". See "
+          "types.db(5) for details.",
+          p->value.type, p->name);
+
+    return;
+  }
+
+  int i = 0;
```

Comment:
This declaration is out of place. Please declare `i` in the loop header, like so:

```c
for (size_t i = 0; i < values_len; ++i) {
```

Since `values_len` is a `size_t`, `i` should be of the same type. Otherwise compilers will complain about comparing signed and unsigned integers.